src/serac/numerics/functional/tensor.hpp

```diff
@@ -1170,58 +1172,147 @@ SERAC_HOST_DEVICE bool is_symmetric_and_positive_definite(tensor<double, 3, 3> A
   };
 
-  tensor<double, n> x{};
+  auto           U = A;
+  auto           L = DenseIdentity<n>();
+  tensor<int, n> P(make_tensor<n>([](auto i) { return i; }));
 
   for (int i = 0; i < n; i++) {
     // Search for maximum in this column
-    double max_val = abs(A[i][i]);
+    double max_val = abs(U[i][i]);
 
     int max_row = i;
     for (int j = i + 1; j < n; j++) {
-      if (abs(A[j][i]) > max_val) {
-        max_val = abs(A[j][i]);
+      if (abs(U[j][i]) > max_val) {
+        max_val = abs(U[j][i]);
         max_row = j;
       }
     }
 
-    swap(b[max_row], b[i]);
-    swap(A[max_row], A[i]);
+    swap(P[max_row], P[i]);
+    swap(U[max_row], U[i]);
+  }
 
-    // zero entries below in this column
+  for (int i = 0; i < n; i++) {
+    // zero entries below in this column in U
+    // and fill in L entries
     for (int j = i + 1; j < n; j++) {
-      double c = -A[j][i] / A[i][i];
-      A[j] += c * A[i];
-      b[j] += c * b[i];
-      A[j][i] = 0;
+      double c = U[j][i] / U[i][i];
+      L[j][i]  = c;
+      U[j] -= c * U[i];
+      U[j][i] = 0;
+    }
+  }
+
+  return tuple{P, L, U};
+}
+
+/**
+ * @brief Solves Ax = b for x using Gaussian elimination with partial pivoting
+ * @param[in] A The coefficient matrix A
+ * @param[in] b The righthand side vector b
+ * @return x The solution vector
+ */
+template <typename T, int n>
+SERAC_HOST_DEVICE constexpr tensor<T, n> linear_solve(const tensor<T, n, n> A, const tensor<T, n>& b)
+{
+  auto const [P, L, U] = lu(A);
+
+  // Forward substitution
+  // solve Ly = b
+  tensor<T, n> y{};
+  for (int i = 0; i < n; i++) {
+    auto c = b[P[i]];
+    for (int j = 0; j < i; j++) {
+      c -= L[i][j] * y[j];
     }
+    y[i] = c / L[i][i];
   }
 
-  // Solve equation Ax=b for an upper triangular matrix A
+  // Back substitution
+  // Solve Ux = y
+  tensor<T, n> x{};
   for (int i = n - 1; i >= 0; i--) {
-    x[i] = b[i] / A[i][i];
-    for (int j = i - 1; j >= 0; j--) {
-      b[j] -= A[j][i] * x[i];
+    auto c = y[i];
+    for (int j = i + 1; j < n; j++) {
+      c -= U[i][j] * x[j];
+    }
+    x[i] = c / U[i][i];
+  }
+
+  return x;
+}
+
+/**
+ * @overload
+ * @note Extends solve to multiple righthand sides
+ * @param[in] A The coefficient matrix A
+ * @param[in] b The matrix of righthand side vectors
+ * @return x The matrix of solution vectors
+ */
+template <typename T, int n, int m>
+SERAC_HOST_DEVICE constexpr tensor<T, n, m> linear_solve(const tensor<T, n, n>& A, const tensor<T, n, m>& b)
+{
+  auto [P, L, U] = lu(A);
+
+  // Forward substitution
+  // solve Ly = b
+  tensor<T, n, m> y{};
+  for (int i = 0; i < n; i++) {
+    auto c = b[P[i]];
+    for (int j = 0; j < i; j++) {
+      c -= L[i][j] * y[j];
     }
+    y[i] = c / L[i][i];
+  }
+
+  // Back substitution
+  // Solve Ux = y
+  tensor<T, n, m> x{};
+  for (int i = n - 1; i >= 0; i--) {
+    auto c = y[i];
+    for (int j = i + 1; j < n; j++) {
+      c -= U[i][j] * x[j];
+    }
+    x[i] = c / U[i][i];
   }
```

Comment:
in theory, the different shapes of the rhs argument can be handled in a single implementation, doing something like
```cpp
template <typename S, typename T, int n, int ... m>
SERAC_HOST_DEVICE constexpr auto linear_solve(const tensor<S, n, n>& A, const tensor<T, n, m ... >& b)
{
  using output_type = decltype(T{} / S{});
  auto [P, L, U] = lu(A);

  // Forward substitution
  // solve Ly = b
  tensor<output_type, n, m ... > y{};
  ...

  // Back substitution
  // Solve Ux = y
  tensor<output_type, n, m> x{};
  ...
  
  return x;
}
```

we might want to try that and see if g++11 handles it correctly or not

Reply:
I changed it and the CI didn't choke. Looks like it works.